### ISSUE TITLE
Getting `Rubinius::ToolSet::Runtime::CompileError` from some of rspec-expectations files

### ISSUE THREAD

**Issue by myronmarston** [rubinius/rubinius#2934]
For RSpec's travis builds we've started getting `Rubinius::ToolSet::Runtime::CompileError` errors for some files from rspec-expectations.  It's inconsistent, though; originally it happened once or twice and appeared to be purely transient because the build would pass on retry, but now it's happening much more consistently.

Note that it varies which file in rspec-expectations triggers the compile error; for example, in [this build](https://travis-ci.org/rspec/rspec-core/jobs/18589997#L325) it's [match.rb](https://github.com/rspec/rspec-expectations/blob/b5afcfbcfb0ca3ddefee1d469aa246ac7125a620/lib/rspec/matchers/built_in/match.rb) but in [this build](https://travis-ci.org/rspec/rspec-core/jobs/18588991#L356) it's [contain_exactly.rb](https://github.com/rspec/rspec-expectations/blob/b5afcfbcfb0ca3ddefee1d469aa246ac7125a620/lib/rspec/matchers/built_in/contain_exactly.rb).

These files are [loaded lazily via autoload](https://github.com/rspec/rspec-expectations/blob/b5afcfbcfb0ca3ddefee1d469aa246ac7125a620/lib/rspec/matchers/built_in.rb) so the failure happens the first time the named matcher is used.  After that, the constant remains undefined so the build gets "Missing or uninitialized constant: ..." `NameError`s for any later use of the matcher that failed to compile.

For both of these specific build failure examples, travis reports this rubinius version:

```
rubinius 2.2.5 (2.1.0 e543ba32 2014-02-08 JI) [x86_64-linux-gnu]
```

**Comment by robin850** [comment]
Hi there!

I don't know if it's due to the same bug (at least we don't have the same exception raised) but we also have a thread-safety issue with autoloading. At the moment, Action Pack's test suite runs in parallel with Minitest's `parallelize_me!` method.

The problem is that most constants are autoloaded and we have a `ActionDispatch::Session` constant (a module) and a `ActionDispatch::Request::Session` constant (a class which is autoloaded). When we are calling a method on the latter within `ActionDispatch::Request`, it is trying to call the method on the former (which doesn't respond to it). Here's a minimal reproduction gist to showcase the problem:

~~~ruby
require 'minitest'
require 'minitest/autorun'

module A
  module C; end

  autoload :B, 'fake_file'

  class B
    autoload :C, 'c'

    def self.bar
      C.foo
    end
  end
end

class B < A::B; end

class FooTest < Minitest::Test
  def self.run_one_method(klass, method_name, reporter)
    Minitest.parallel_executor << [klass, method_name, reporter]
  end

  def test_b
    B.bar
  end
end
~~~

The `self.run_one_method` is pasted from Minitest ; it mimics a `paralellize_me!` call which will run tests inside separate threads. Here `fake_file.rb` is just an empty file and the `c.rb` file contains:

~~~ruby
module A
  class B
    class C
      def self.foo; end
    end
  end
end
~~~

The interesting thing is that if we remove the `autoload` call on `B` or if we simply write the definition of `C` like `class A::B::C`, the problem goes away. Hoping it can help to track down the problem here.

**Comment by YorickPeterse** [comment]
@robin850 Hm, interesting. In my own experience autoload is definitely not thread-safe, even if we might claim otherwise.

**Comment by robin850** [comment]
@YorickPeterse : Yep, it looks like. I don't know whether it could be useful but here's a more "ruby-ish" example that doesn't rely on Minitest to showcase the problem: https://gist.github.com/robin850/3eeaed2538f50a9887c2 ; you just need to run `ruby -I"." foo.rb` once the gist is cloned.

**Comment by jemc** [comment]
@robin850 - can you verify that your repro still fails on rbx-master and open a new issue with a link to your repro so we can track work there?

**Comment by YorickPeterse** [comment]
The script @robin850 wrote still reproduces for me using rbx master.

**Comment by YorickPeterse** [comment]
In commit https://github.com/rubinius/rubinius/commit/b57399f8f44c9afbe5b2785abd85943a8f46ddd3 I made some changes to hopefully finally make autoloading fully thread-safe (famous last words). @myronmarston and @robin850: could you try to run your scripts using said commit and see if the issue still persists?

**Comment by YorickPeterse** [comment]
Hm, apparently the script discussed in https://github.com/rubinius/rubinius/issues/2934#issuecomment-36667492 still triggers a constant error. It appears this hydra grew a few extra heads.

**Comment by myronmarston** [comment]
I would try if I had such a script but I was never able to repro locally.  The errors I saw only happened on travis.

**Comment by headius** [comment]
As far as I can tell, RSpec::Matchers::Builtin contains no autoloads, and that's the constant that's going missing. I can even remove the last element and it still fails on JRuby and Rubinius. I'm not sure the script from https://github.com/rubinius/rubinius/issues/2934#issuecomment-36667492 is actually hitting any autoload-related bugs.

**Comment by headius** [comment]
I'm not sure if it explains the original issue, which is in a deeper namespace, but the logic for autoloading RSpec::Matcher is not thread-safe. See https://github.com/rspec/rspec-dev/issues/121. It may not explain the original issue, but it does relieve Rubinius of any culpability in the RSpec::Matchers::BuiltIn error.

**Comment by YorickPeterse** [comment]
@headius Hm, interesting. I spent some time looking at Rubinius' `find_const_fast` instruction (used to retrieve the `RSpec` constant in the above script) but couldn't figure out what could be causing this error in said instruction. I hadn't realized RSpec uses a custom `const_get`.

**Comment by myronmarston** [comment]
> I hadn't realized RSpec uses a custom const_get.

Yeah, it would be nice not to, but we haven't been able to come up with an alternate solution that both allows users to configure RSpec to use an alternates to rspec-expectations/rspec-mocks, while still making it "just work" for users who use the default stack.  See my response on rspec/rspec-dev#121 for more detail.

If you want to test the `RSpec::Matchers::BuiltIn` autoloads like you originally intended to, I think you can just change the require in your script:

``` ruby
require 'rspec/matchers'

threads = []

100.times do
  threads << Thread.new { sleep(rand); RSpec::Matchers::BuiltIn::Match }
end

threads.each(&:join)
```

That'll avoid the `const_missing`/`const_get` piece @headius pointed out.

> It may not explain the original issue, but it does relieve Rubinius of any culpability in the RSpec::Matchers::BuiltIn error.

Yep, I don't think it explains the `Rubinius::ToolSet::Runtime::CompileError` error I was originally reporting.  In that case, the error is happening long after `rspec/matchers` has been loaded, when one of the `RSpec::Matchers::BuiltIn` constants is accessed for the first time.

**Comment by robin850** [comment]
> could you try to run your scripts using said commit and see if the issue still persists?

With current master (as of fbb3f1e4), the given scripts are still failing unfortunately.

**Comment by ahmadsherif** [comment]
After some digging, it may be not related to `autoload`. In one of the [failing examples](https://github.com/rspec/rspec-core/blob/facc88c0ae94836d5bc65e5a5f0a49d36d2c3de6/spec/rspec/core/drb_options_spec.rb#L9) `File.exist?` is stubbed to return `false`, the parser [uses this method](https://github.com/rubinius/rubinius-code/blob/fdf2230b1458a76305b3d9fb828b80add89de58b/lib/rubinius/code/melbourne.rb#L57) to check for file existence, so it fails. I'm not sure what Rubinius can do for such cases.

**Comment by brixen** [comment]
If I'm understanding the thread here correctly, this is a concurrency issue in RSpec (see referenced issue https://github.com/rspec/rspec-dev/issues/121).
